### PR TITLE
System Configuration helper methods & tests for Port validation and i…

### PR DESCRIPTION
**Pull request by NateThornton**
System Configuration helper methods & tests for Port validation and iteration

### CODE REVIEW

**Comment by roehrich-hpe** [comment]
What does 'make manifests' do with the rename of that `_types.go` file?

This helper thing and its test should be in a `utils/` dir,  rather than in `api/`.
Things like this become more obvious when we start rolling the CRD versions...and we have to copy this stuff to the next version.

**Comment by NateThornton** [comment]
Running `make manifests` makes no changes.

I can't move these methods into utils since they are class member functions. Besides, I prefer this file layout since it makes importing easier (why import both v1alpha1 and utils?), and it's easier to understand (where tests are in the same directory as the definitions).

We can re-arrange these files if a new version is ever introduced. But I don't want to sacrifice a good logical structure and ownership for an unknown future that might never happen.

File: api/v1alpha1/system_configuration_test.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2021, 2022 Hewlett Packard Enterprise Development LP
```
Comment (bdevcich-hpe) [inline api/v1alpha1/system_configuration_test.go:2]: 2023

File: api/v1alpha1/system_configuration_test.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+			Expect(config.Validate() == nil).To(Equal(isValid))
+		},
+		// Integer values
+		Entry("Valid integer", "1", true),
+		Entry("Zero", "0", false),
+		Entry("Negative", "-1", false),
+		Entry("Funky whitespace", "  1  ", false),
+		Entry("Too small", "0", false),
+		Entry("Too large", "65536", false),
```
Comment (bdevcich-hpe) [inline api/v1alpha1/system_configuration_test.go:45]: Add a test for 65535

File: api/v1alpha1/system_configuration_test.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+		// String values
+		Entry("Valid port range", "1-2", true),
+		Entry("Invalid port range", "1-", false),
+		Entry("Funky whitespace", " 1 -  2", false),
+		Entry("Start too small", "0-1", false),
+		Entry("Start too large", "65536-65537", false),
+		Entry("End too small", "1-0", false),
+		Entry("End too large", "1-65536", false),
```
Comment (bdevcich-hpe) [inline api/v1alpha1/system_configuration_test.go:54]: Same here. Do all the ports 1-65535

**Comment by bdevcich-hpe** [comment]
> This helper thing and its test should be in a `utils/` dir, rather than in `api/`.

It _feels_ to me that this is the right approach, but I'm fine with it either way.

However, if we move this to utils, how do we ensure that it's backwards compatible with v1 if we move to v2? It seems like having to copy it wouldn't be too bad if left in api. However, I have no experience with k8s versioning.

**Comment by roehrich-hpe** [comment]
> However, if we move this to utils, how do we ensure that it's backwards compatible with v1 if we move to v2? It seems like having to copy it wouldn't be too bad if left in api.

That's a good point.  Consumers of the dws api should be allowed to use older versions of the CRD, so they'll need the libraries that are known to match that older version.

So I will retract my suggestion.  Let's let each version of the dws api to have its own copy of this library.

**Comment by NateThornton** [comment]
I'm actually coming around to the idea of moving this to utils. Most of the code operates on a `Ports []intstr.IntOrString` and is really independent of any system configuration so long as it carries a `Ports` value of the same type.

File: utils/ports/ports_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func TestAPIs(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	RunSpecs(t, "Webhook Suite")
```
Comment (roehrich-hpe) [inline utils/ports/ports_test.go:33]: Some other text here.

**Review by roehrich-hpe — APPROVED** [review]

**Pull request merged**